File: chapter4 - Trees and Graphs/4.4_checkBalanced/Java/inefficient/CheckBalanced.java
package inefficient;

public class CheckBalanced {
  public int getHeight(TreeNode root) {
    if (root == null) return -1; // Base case
    return Math.max(getHeight(root.left), getHeight(root.right)) + 1;
  }

  public boolean isBalanced(TreeNode root) {
    if (root == null) return true; // Base case

    int heightDiff = getHeight(root.left) - getHeight(root.right);

    if (Math.abs(heightDiff) > 1) {
      return false;
    } else { // Recurse
      return isBalanced(root.left) && isBalanced(root.right);
    }
  }
}

File: chapter4 - Trees and Graphs/4.4_checkBalanced/JavaScript/CheckBalanced.js
function checkBalanced(root) {
  if (!root) {
    return false;
  }

  function helperFunc(root) {
    if (!root) {
      return 0;
    }

    const leftResult = helperFunc(root.left);
    const rightResult = helperFunc(root.right);
    const total = Math.abs(leftResult - rightResult);

    if (leftResult < 0 || rightResult < 0 || total > 1) {
      return -1;
    } else {
      return 1 + Math.max(leftResult, rightResult);
    }
  };

  const leftResult = helperFunc(root.left);
  const rightResult = helperFunc(root.right);
  const total = Math.abs(leftResult - rightResult);

  if (leftResult < 0 || rightResult < 0 || total > 1) {
    return false
  } else {
    return true;
  }
}

module.exports = checkBalanced;